druid-shell/src/platform/x11/window.rs

```diff
@@ -615,6 +788,171 @@ impl Window {
+            .map(|p| p.waiting_on.is_some())
+            .unwrap_or(false))
+    }
+
+    fn set_needs_present(&self, val: bool) -> Result<(), Error> {
+        if let Some(present) = borrow_mut!(self.present_data)?.as_mut() {
+            present.needs_present = val;
+        }
+        Ok(())
+    }
+
+    fn needs_present(&self) -> Result<bool, Error> {
+        Ok(borrow!(self.present_data)?
+            .as_ref()
+            .map(|p| p.needs_present)
+            .unwrap_or(false))
+    }
+
+    /// Frees all resources used for the Present extension, and falls back to just using EXPOSE.
+    pub fn disable_present(&self) -> Result<(), Error> {
+        if let Some(present) = borrow_mut!(self.present_data)?.as_mut() {
+            present.destroy_x_resources(self.app.connection());
+        }
+        *borrow_mut!(self.present_data)? = None;
+        Ok(())
+    }
+}
+
+impl PresentData {
+    /// Frees all the X resources that we hold. Note the potential for memory leaks, because this
+    /// is not called automatically on drop. (Maybe we should add an `Rc<Connection>` to
+    /// `PresentData` so that we can do it on drop?)
+    fn destroy_x_resources(&mut self, conn: &Rc<Connection>) {
+        self.free_pixmaps(conn);
+        xcb::xfixes::destroy_region(conn, self.region);
+    }
+
+    /// Frees all the X pixmaps that we hold.
+    fn free_pixmaps(&mut self, conn: &Rc<Connection>) {
+        for &p in &self.all_pixmaps {
+            xcb::free_pixmap(conn, p);
+        }
+        self.all_pixmaps.clear();
+        self.idle_pixmaps.clear();
+    }
+
+    fn resize(
+        &mut self,
+        conn: &Rc<Connection>,
+        window_id: u32,
+        width: u16,
+        height: u16,
+        depth: u8,
+    ) {
+        // TODO: we can avoid reallocating so frequently by allowing "oversize" pixmaps.
+        if (width, height, depth) != (self.width, self.height, self.depth) {
+            self.free_pixmaps(conn);
+            self.width = width;
+            self.height = height;
+            self.depth = depth;
+            self.create_pixmaps(conn, window_id);
+        }
+    }
+
+    /// Creates a new pixmap for rendering to. The new pixmap will be first in line for rendering.
+    fn create_pixmap(&mut self, conn: &Rc<Connection>, window_id: u32) -> Pixmap {
+        let pixmap_id = conn.generate_id();
+        xcb::create_pixmap(
+            &conn,
+            self.depth,
+            pixmap_id,
+            window_id,
+            self.width,
+            self.height,
+        );
+        self.all_pixmaps.push(pixmap_id);
+        self.idle_pixmaps.push(pixmap_id);
+        pixmap_id
+    }
+
+    fn create_pixmaps(&mut self, conn: &Rc<Connection>, window_id: u32) {
+        if !self.all_pixmaps.is_empty() {
+            log::error!("BUG: need to free the pixmaps before creating new ones");
+            self.free_pixmaps(conn);
+        }
+
+        // Two buffers is probably enough. We'll create more on demand.
```

Comment:
Under what circumstances do we need more than two? :thinking:

Reply:
As far as I can tell, X doesn't give any guarantees about when it will give us the buffer back. In my testing, I basically always get it back just after vsync; since we present once per vsync, we just alternate between buffers.

One latency-reducing possibility for the future is to render more frequently when we're getting input: if rendering is fast and we get input sometime during the inter-frame time, we could try rendering in the hope that we'll beat the vsync deadline. We might need an extra buffer for this.

Reply:
Okay, then my question would be - are these additional buffers additive or replacements?
So if we have say 10 buffers and we instruct X to render all of these in a row. Does only the last one get rendered, or does this manifest as a 10 frame delay?

If it's extra delay, then we should attempt an optimization solution. If just the last one gets used, then we should add a comment stating that these extra buffers wont cause actual delays.